### ISSUE TITLE
TypeScript error "implicitly has an 'any' type" with @types/react >= 16.4.8

### ISSUE THREAD

**Issue by benneq** [jaredpalmer/formik#881]
## Current Behavior

I get the following error message by TypeScript compiler (VSCode doesn't complain) on the `onSubmit` method within the `<Formik ... />` component:
```
Parameter 'values' implicitly has an 'any' type.
```

## Steps to Reproduce

Use `formik` in version `1.1.0` - `1.2.0` (no older version tested)
Use `@types/react` in version `16.4.8` - `16.4.13` (tested all versions inbetween, too) 
```
import { Form, Field, Formik } from 'formik';
export const MyComponent = (props) => {
  return {
    <Formik
      initialValues={{email:'', password:''}}
      onSubmit={(values) => console.log(values.email, values.password)}
      render={() =>  ... }
    />
  }
}
```

## Expected behavior

No error message.

## Suggested solution(s)

I'm not sure what exactly the problem is. But if I add generic parameter `any` to the `<Formik` component, it works. But then of course my `onSubmit` `values` is of type `any`:
```
<Formik<any> ... />
```

## Additional context

I'm currently still using `@types/react` `16.4.7` because that's the latest version that doesn't cause this error.
As said above, I've tested some different versions of the last weeks, and the error only happens when I switch to `@types/react` >= `16.4.8`.
And also VSCode shows the type of `values` correct:
```
(parameter) values: {
  email: string;
  password: string;
}
```
The error only happens when compiling and watching the TypeScript project.

**My tsconfig.json:**
```
{
  "compilerOptions": {
    "baseUrl": ".",
    "outDir": "build/dist",
    "module": "esnext",
    "target": "es5",
    "lib": ["es6", "es2017", "dom", "esnext.asynciterable"],
    "sourceMap": true,
    "jsx": "react",
    "moduleResolution": "node",
    "rootDir": "src",
    "forceConsistentCasingInFileNames": true,
    "noImplicitAny": true,
    "noImplicitThis": true,
    "alwaysStrict": true,
    "strictNullChecks": true,
    "noImplicitReturns": true,
    "noUnusedLocals": false,
    "noFallthroughCasesInSwitch": true
  },
  "exclude": [
    "node_modules",
    "build",
    "scripts",
    "acceptance-tests",
    "webpack",
    "jest",
    "src/setupTests.ts"
  ]
}
```

**CodeSandbox Link**: https://codesandbox.io/s/jzo3lxw3nv
Not really useful. Does show no error. But also doesn't allow to set the generic parameter of the `<Formik<...> ... />` component.

---

- Formik Version: 1.2.0 (also tested 1.1.0, 1.1.1, 1.1.2)
- React Version: 16.5.0 (also tested 16.4.x)
- TypeScript Version: 3.0.3
- Browser and Version: Chrome 69
- OS: macOS High Sierra 10.13.6
- Node Version: 10.9.0
- Package Manager and Version: npm 6.2.0

**Comment by wilcoschoneveld** [comment]
I have the same issue, thanks for the write-up!

**Comment by wilcoschoneveld** [comment]
You can provide the value typing directly to the formik component instead of `any`:
```
<Formik<{ name: string }>
  initialValues={{
    name: ""
  }}
  ....
/>
```

This works with `@types/react` `^16.4.13`

**Comment by benneq** [comment]
I already said that above (using `<any>`). But that's no solution. I don't want to specify the types twice. 

Of course there's another workaround:
```
const val = { name: "" }
...
<Formik<typeof val>
  initialValues={val}
  ...
/> 
```

But I don't want a workaround. I just want the old behavior :D

**Comment by wilcoschoneveld** [comment]
Yeah it also is bugging me a bit, defining types twice. But I can also see the case where the type of a value is a union of types (e.g. `category: "cats" | "dogs"`), which can not be derived from just an initialValue. Therefore maybe it's not so bad to define the type explicitly.

**Comment by jaredpalmer** [comment]
Realistically you always need a type, but i changed the default value of the generic from `object` to `{}` which may not be desired i guess. I have learned the hard way not to update our @types/react very often (_never_ cough cough) so I apologize. One of the terrible flaws of definitiely typed is that it is community driven and we can't always control what gets merged.

**Comment by benneq** [comment]
Maybe you could change it back to `object` with the next version or use generics?

Using an old version of `@types/react` isn't always possible. When React 17 gets released the current react types will make huge problems. So for now it's fine to use an old version, but in the future it will hurt.

**Comment by jaredpalmer** [comment]
Agree. Will change it back to object.

**Comment by benneq** [comment]
Any progress?

Now I'm at the point where I can't update several other libs (for example react-apollo), because they depend on the new @types/react version.

**Comment by jaredpalmer** [comment]
Not yet. I’m trying to get 3.1 working, but now there is a different problem. See #898

**Comment by benneq** [comment]
Seems to be fixed with Formik 1.3.1 . Intented or by accident?

**Comment by jaredpalmer** [comment]
Intended
